awx/main/models/notifications.py

```diff
@@ -201,56 +247,145 @@ def get_absolute_url(self, request=None):
+                                                {'labels': ['count', 'results']}]}]
+
+    @classmethod
+    def context_stub(cls):
+        """Returns a stub context that can be used for validating notification messages.
+        Context has the same structure as the context that will actually be used to render
+        a notification message."""
+        context = {'job': {},
+                   'job_friendly_name': 'stub',
+                   'url': 'stub'}
+
+        def build_context_for_fields(node, fields):
+            for f in fields:
+                if type(f) is dict:
+                    subname, subfields = f.copy().popitem()
+                    node[subname] = {}
+                    build_context_for_fields(node[subname], subfields)
+                else:
+                    node[f] = 'stub'
+        build_context_for_fields(context['job'], cls.JOB_FIELDS_WHITELIST)
+
+        return context
+
+    def context(self, serialized_job):
+        """Returns a context that can be used for rendering notification messages.
+        Context contains whitelisted content retrieved from a serialized job object
+        (see JobNotificationMixin.JOB_FIELDS_WHITELIST), the job's friendly name,
+        and a url to the job run."""
+        context = {'job': {},
+                   'job_friendly_name': self.get_notification_friendly_name(),
+                   'url': self.get_ui_url()}
+
+        def build_context(node, fields, whitelisted_fields):
+            for safe_field in whitelisted_fields:
+                if type(safe_field) is dict:
+                    field, whitelist_subnode = safe_field.copy().popitem()
+                    # ensure content present in job serialization
+                    if field not in fields:
+                        continue
+                    subnode = fields[field]
+                    node[field] = {}
+                    build_context(node[field], subnode, whitelist_subnode)
+                else:
+                    # ensure content present in job serialization
+                    if safe_field not in fields:
+                        continue
+                    node[safe_field] = fields[safe_field]
+        build_context(context['job'], serialized_job, self.JOB_FIELDS_WHITELIST)
+
+        return context
+
     def get_notification_templates(self):
         raise RuntimeError("Define me")
 
     def get_notification_friendly_name(self):
         raise RuntimeError("Define me")
 
-    def _build_notification_message(self, status_str):
+    def notification_data(self):
+        raise RuntimeError("Define me")
+
+    def build_notification_message(self, nt, status):
+        env = sandbox.ImmutableSandboxedEnvironment()  # TODO: sub-class to define unsafe attrs
+
+        from awx.api.serializers import UnifiedJobSerializer  # Need to import here (To avoid circular import?)
+        job_serialization = UnifiedJobSerializer(self).to_representation(self)
```

Comment:
This is pretty darn cool:

![image](https://user-images.githubusercontent.com/214912/62067470-d339c200-b201-11e9-8dd7-56ae5fa11c8b.png)

Reply:
Workflows work too, nice:
![image](https://user-images.githubusercontent.com/214912/62068106-5e678780-b203-11e9-9eb3-315fd018eb50.png)

Reply:
Yeah, it was super-handy to have this when building the Jinja context for notification messages / templates